### PR TITLE
Compute relative_path of pages using PathManager

### PR DESCRIPTION
**Pull request by ashmaroli**
- This is a 🔨 code refactoring change.
- I've added tests.

## Summary

`File.join` is sensitive to `nil` arguments. So we need to coerce arguments to String to be safer.
Using `Jekyll::PathManager.join` on the other hand allows optimized execution for same results.

## Context

Prior discussions: https://github.com/jekyll/jekyll/pull/8404#discussion_r494402547

### CODE REVIEW

**Comment by ashmaroli** [comment]
## Benchmark report

Results from Ruby 2.4 on Windows

```
Warming up --------------------------------------
       nil_page slow    23.239k i/100ms
       nil_page fast    17.135k i/100ms
Calculating -------------------------------------
       nil_page slow    250.293k (A±16.1%) i/s -      1.232M in   5.016589s
       nil_page fast    166.101k (A± 5.3%) i/s -    839.615k in   5.067538s

Comparison:
       nil_page slow:   250293.3 i/s
       nil_page fast:   166101.0 i/s - 1.51x  (A± 0.00) slower

Warming up --------------------------------------
     empty_page slow    26.692k i/100ms
     empty_page fast    16.557k i/100ms
Calculating -------------------------------------
     empty_page slow    295.685k (A± 5.0%) i/s -      1.495M in   5.067854s
     empty_page fast    183.103k (A± 5.0%) i/s -    927.192k in   5.075815s

Comparison:
     empty_page slow:   295684.9 i/s
     empty_page fast:   183102.7 i/s - 1.61x  (A± 0.00) slower

Warming up --------------------------------------
      root_page slow    18.165k i/100ms
      root_page fast    14.024k i/100ms
Calculating -------------------------------------
      root_page slow    191.299k (A±11.2%) i/s -    962.745k in   5.081186s
      root_page fast    184.917k (A± 5.6%) i/s -    925.584k in   5.022202s

Comparison:
      root_page slow:   191299.4 i/s
      root_page fast:   184916.8 i/s - same-ish: difference falls within error

Warming up --------------------------------------
    nested_page slow    16.210k i/100ms
    nested_page fast    29.264k i/100ms
Calculating -------------------------------------
    nested_page slow    159.222k (A± 5.1%) i/s -    810.500k in   5.103311s
    nested_page fast    297.392k (A± 6.9%) i/s -      1.492M in   5.042992s

Comparison:
    nested_page fast:   297392.0 i/s
    nested_page slow:   159222.4 i/s - 1.87x  (A± 0.00) slower

Warming up --------------------------------------
   slashed_page slow    12.482k i/100ms
   slashed_page fast    18.147k i/100ms
Calculating -------------------------------------
   slashed_page slow    122.108k (A± 4.6%) i/s -    611.618k in   5.019493s
   slashed_page fast    189.240k (A±12.1%) i/s -    943.644k in   5.046522s

Comparison:
   slashed_page fast:   189239.8 i/s
   slashed_page slow:   122107.6 i/s - 1.55x  (A± 0.00) slower
```

The proposed change is only slow for edge-cases. However, there is significant improvement for pages in subdirectories.

File: test/test_page_without_a_file.rb [hunk 1]
```diff
@@ -28,6 +34,44 @@ def render_and_write
...
+    should "have non-frozen path and relative_path attributes" do
+      page = PageWithoutAFile.new(@site, @site.source, "foo", "bar.md")
+      assert_equal "foo", page.instance_variable_get(:@dir)
+      assert_equal "bar.md", page.instance_variable_get(:@name)
+      assert_equal "foo/bar.md", page.path
+      assert_equal "foo/bar.md", page.relative_path
+      refute page.relative_path.frozen?
+
+      # reset attributes
+      rehash_page!(page, :dir => nil, :name => nil)
```
Comment (parkr) [inline test/test_page_without_a_file.rb:46]: Instead of doing this, why don't we just create a new PageWithoutAFile instance? I don't think the tests knowing about the internal instance variables is a great idea. Also, These blocks could be separate Tests to help describe what you're testing for.

Comment (ashmaroli) [reply test/test_page_without_a_file.rb:46]: Good question, @parkr.
The reason we can't create a new `PageWithoutAFile` instance is because of the following lines:
https://github.com/jekyll/jekyll/blob/fe898af92fadecbec6b4b2ec69c1cb2d85eb46e2/lib/jekyll/page.rb#L44-L48

`Jekyll::PageWithoutAFile.new(site, site.source, nil, nil)` fails as `Jekyll.sanitized_path` doesn't cast its arguments into strings automatically.

Comment (ashmaroli) [reply test/test_page_without_a_file.rb]: `Jekyll.sanitized_path` has been patched via https://github.com/jekyll/jekyll/commit/98b7a4626a86de4e068ffa6bd927f9437a444c28

**Review by DirtyF — APPROVED** [review]
Looks good, I'm not sure how to fix the JRuby compilation error.

File: test/test_page_without_a_file.rb [hunk 2]
```diff
@@ -28,6 +28,21 @@ def render_and_write
...
+        ["foo", "bar.md"]     => "foo/bar.md",
+        [nil, nil]            => "",
+        ["", ""]              => "",
+        ["/lorem/", "/ipsum"] => "lorem/ipsum",
+        %w(lorem ipsum)       => "lorem/ipsum",
+      }.each do |(dir, name), result|
+        page = PageWithoutAFile.new(@site, @site.source, dir, name)
+        assert_equal result, page.path
+        assert_equal result, page.relative_path
+        refute page.relative_path.frozen?
```
Comment (parkr) [inline test/test_page_without_a_file.rb:42]: Is the relative path modified? Why is it important that it's unfrozen?

**Review by parkr — APPROVED** [review]
Nice! One question.

File: test/test_page_without_a_file.rb (hunk 2 above)
Comment (ashmaroli) [reply test/test_page_without_a_file.rb:42]: For backwards-compatibility. It is an unfrozen string in v4.1.
`Jekyll::Pathfinder.join` returns a frozen cached string. So this test confirms that `page.relative_path` hasn't changed behavior in v4.2..

**Comment by ashmaroli** [comment]
Thank you the reviews @DirtyF  and @parkr 
@jekyllbot: merge +fix
